### PR TITLE
drivers: adc: adc_sam: Introduce Atmel SAM ADC driver

### PR DESCRIPTION
**Pull request by pdgendt**
Introduce Atmel SAM ADC driver, and add it to the Atmel SAM4S SoC. 
All of the timings are currently set for the controller instead of individual channels, so only `ADC_ACQ_TIME_DEFAULT` is supported.

Tested on custom board.

@nandojve I will probably need to add this to the `sam4s_xplained` board to take it along in CI. Any suggestions to make it testable?

### CODE REVIEW

**Comment by nandojve** [comment]
> Introduce Atmel SAM ADC driver, and add it to the Atmel SAM4S SoC. All of the timings are currently set for the controller instead of individual channels, so only `ADC_ACQ_TIME_DEFAULT` is supported.
> 
> Tested on custom board.
> 
> @nandojve I will probably need to add this to the `sam4s_xplained` board to take it along in CI. Any suggestions to make it testable?

Hi @pdgendt,

You can make PB0 and PB1 available as ADC inputs since this is instructed on the SAM4S_Xplained schematic.
About CI, it will be good if we can get both sample and tests running [adc sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/adc) + [adc tests](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/adc).

**Comment by pdgendt** [comment]
@nandojve I've updated the boards/samples/tests

**Comment by nandojve** [comment]
> @nandojve I've updated the boards/samples/tests

Hi @pdgendt ,

This still as a draft or you want to have it on v3.3?

**Comment by pdgendt** [comment]
> This still as a draft or you want to have it on v3.3?

Hi @nandojve, I'll clean it up to get it in v3.3.

**Comment by pdgendt** [comment]
Hi @nandojve, would you be able to test using the `sam4s_xplained` board?

**Comment by MaureenHelm** [comment]
@nandojve ping

**Comment by nandojve** [comment]
Hi @pdgendt ,

I configured sample to run as single end instead diff on channel 5 (PB1) and I didn't change temperature configuration.
It seems that there are something not proper with the driver. The read values oscillates and channel 5 not read correct.

The same HW configuration using ASF->SAM->Drivers->ADC sample show different behavior.

```diff
+++ b/samples/drivers/adc/boards/sam4s_xplained.overlay
@@ -9,7 +9,7 @@
 / {
        zephyr,user {
                /* adjust channel number according to pinmux in board.dts */
-               io-channels = <&adc0 2>, <&adc0 15>;
+               io-channels = <&adc0 5>, <&adc0 15>;
        };
 };
 
@@ -17,14 +17,13 @@
        #address-cells = <1>;
        #size-cells = <0>;
 
-       /* External ADC(+)/ADC(-) differential inputs */
-       channel@2 {
-               reg = <2>;
+       /* External ADC(+) */
+       channel@5 {
+               reg = <5>;
                zephyr,gain = "ADC_GAIN_1";
                zephyr,reference = "ADC_REF_EXTERNAL0";
                zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
-               zephyr,input-positive = <4>;
-               zephyr,input-negative = <5>;
+               zephyr,input-positive = <5>;
                zephyr,resolution = <12>;
        };

```

```
*** Booting Zephyr OS build v3.3.0-rc2-204-g86548f258c98 ***
ADC reading:
- adc@40038000, channel 5: 64 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1716 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1720 = 0 mV
ADC reading:
- adc@40038000, channel 5: 60 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 62 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 61 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 61 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1726 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1724 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 62 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 62 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 63 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 43 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 1754 = 0 mV
ADC reading:
- adc@40038000, channel 5: 39 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 67 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 0 = 0 mV
- adc@40038000, channel 15: 2047 = 0 mV
ADC reading:
- adc@40038000, channel 5: 146 = 0 mV
- adc@40038000, channel 15: 1717 = 0 mV
ADC reading:
- adc@40038000, channel 5: 1841 = 0 mV
- adc@40038000, channel 15: 1723 = 0 mV
ADC reading:                                                                       
- adc@40038000, channel 5: 3734 = 0 mV                                             
- adc@40038000, channel 15: 2047 = 0 mV
```

**Comment by teburd** [comment]
@pdgendt please take a look

**Comment by nandojve** [comment]
Hi @pdgendt ,

Any new driver should go after #55655.

**Comment by pdgendt** [comment]
> Any new driver should go after #55655.

Hi @nandojve, I will update the PR once merged, for the test on `sam4s_xplained`, I think it's a clocking setting that should be tweaked.

Could you test again with the `prescaler` property set to the maximum `255`? I've done the same on my custom board and the internal temperature sensor is stable.

I don't know what sensible values would be for the timings on the `sam4s_xplained` board, any suggestions?

**Comment by nandojve** [comment]
Hi @pdgendt ,

Thank you for look on it. I'll have time to look on HW at weekend. Right now my reference have been the example inside [Atmel ASF](https://ww1.microchip.com/downloads/en/DeviceDoc/asf-standalone-archive-3.52.0.113.zip) at ASF->SAM->Drivers->ADC.

**Comment by pdgendt** [comment]
> Thank you for look on it. I'll have time to look on HW at weekend. Right now my reference have been the example inside [Atmel ASF](https://ww1.microchip.com/downloads/en/DeviceDoc/asf-standalone-archive-3.52.0.113.zip) at ASF->SAM->Drivers->ADC.

Great, thanks.

I used the example as a reference for my implementation, setting the `prescaler` did yield stable readings, so I guess the driver is functional.

Maybe the bindings could improve, I currently set these values as they would be set in the `MR` register. I could, for example, change the `prescaler` binding to `frequency` binding and calculate the value using your implementation for the peripheral clock.

File: boards/arm/sam4s_xplained/sam4s_xplained.dts [hunk 1]
```diff
@@ -176,6 +176,18 @@
...
 };
 
+&adc0 {
+	status = "okay";
+
+	pinctrl-0 = <&adc0_default>;
+	pinctrl-names = "default";
+
+	prescaler = <2>;
+	startup-time = <1>;
```
Comment (nandojve) [inline boards/arm/sam4s_xplained/sam4s_xplained.dts:186]: startup-time = <4>;

File: boards/arm/sam4s_xplained/sam4s_xplained.dts [hunk 2]
```diff
@@ -176,6 +176,18 @@
...
+	settling-time = <0>;
+	tracking-time = <2>;
```
Comment (nandojve) [inline boards/arm/sam4s_xplained/sam4s_xplained.dts:188]: tracking-time = <1>;

File: drivers/adc/adc_sam.c [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+	/* Reset PDC transfer */
+	adc->ADC_PTCR = ADC_PTCR_RXTDIS | ADC_PTCR_TXTDIS;
+	adc->ADC_RCR = 0U;
+	adc->ADC_RNCR = 0U;
+
+	/* Set prescaler, timings and allow different analog settings for each channel */
+	adc->ADC_MR = ADC_MR_PRESCAL(cfg->prescaler)
+		    | ADC_MR_STARTUP(cfg->startup_time)
+		    | ADC_MR_SETTLING(cfg->settling_time)
+		    | ADC_MR_TRACKTIM(cfg->tracking_time)
```
Comment (nandojve) [inline drivers/adc/adc_sam.c:303]:
```diff
+                   | ADC_MR_TRANSFER(2)
```

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @pdgendt ,

I still not get good results. I don't know why but potentiometer oscillates which not happen with Atmel sample code, maybe a concurrency problem?

File: drivers/adc/adc_sam.c [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+	/* Reset current gain */
+	adc->ADC_CGR &= ~(ADC_CGR_GAIN0_Msk << (channel_id * 2U));
+
+	switch (channel_cfg->gain) {
+	case ADC_GAIN_1_2:
+		if (!channel_cfg->differential) {
+			LOG_ERR("ADC 1/2x gain only allowed for differential channel");
+			return -EINVAL;
+		}
+		/* NOP */
```
[the remark this answers is not in the document]
Comment (pdgendt) [reply drivers/adc/adc_sam.c:117]: 0.5 gain is the default value after resetting, which is done before the switch/case with:

```c
	adc->ADC_CGR &= ~(ADC_CGR_GAIN0_Msk << (channel_id * 2U));
```

File: dts/bindings/adc/atmel,sam-adc.yaml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+  peripheral-id:
+    type: int
+    description: peripheral ID
+    required: true
+
+  prescaler:
+    type: int
+    required: true
+    description: CPU clock prescaler applied to get the ADC clock.
```
[the remark this answers is not in the document]
Comment (pdgendt) [reply dts/bindings/adc/atmel,sam-adc.yaml:25]: The prescaler applies to all channels, can we use the `channel_cfg->acquisition_time` that applies to a single channel?

**Comment by pdgendt** [comment]
Hi @nandojve,

Did you test using `prescaler = <10>`? This would yield an ADC frequency of 5,4 MHz (`120 / 2 / (prescaler + 1`). Which is higher than the allowed maximum of 1 MHz.

The minimum value is `prescaler = <59>` if I applied the correct formula (`120 / 2 / (59 + 1) = 1`).

File: drivers/adc/adc_sam.c [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+	adc->ADC_PTCR = ADC_PTCR_RXTDIS | ADC_PTCR_TXTDIS;
+	adc->ADC_RCR = 0U;
+	adc->ADC_RNCR = 0U;
+
+	/* Set prescaler, timings and allow different analog settings for each channel */
+	adc->ADC_MR = ADC_MR_PRESCAL(cfg->prescaler)
+		    | ADC_MR_STARTUP(cfg->startup_time)
+		    | ADC_MR_SETTLING(cfg->settling_time)
+		    | ADC_MR_TRACKTIM(cfg->tracking_time)
+		    | ADC_MR_ANACH_ALLOWED;
```
[the remark this answers is not in the document]
Comment (pdgendt) [reply drivers/adc/adc_sam.c:304]: See comment https://github.com/zephyrproject-rtos/zephyr/pull/51885#discussion_r1141749364

File: dts/bindings/adc/atmel,sam-adc.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (nandojve) [reply dts/bindings/adc/atmel,sam-adc.yaml:23]: Let's do this in another PR.

**Review by nandojve — APPROVED** [review]
LGTM!

**Review by teburd — APPROVED** [review]

**Pull request merged**